File: app/src/main/java/com/vondear/tools/adapter/AdapterContactCity.java
package com.vondear.tools.adapter;


import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.vondear.tools.R;
import com.vondear.tools.model.ModelContactCity;

import java.util.List;

import cc.solart.turbo.BaseTurboAdapter;
import cc.solart.turbo.BaseViewHolder;


/**
 * @author vondear
 */
public class AdapterContactCity extends BaseTurboAdapter<ModelContactCity, BaseViewHolder> {

    public AdapterContactCity(Context context) {
        super(context);
    }

    public AdapterContactCity(Context context, List<ModelContactCity> data) {
        super(context, data);
    }

    @Override
    protected int getDefItemViewType(int position) {
        ModelContactCity city = getItem(position);
        return city.type;
    }

    @Override
    protected BaseViewHolder onCreateDefViewHolder(ViewGroup parent, int viewType) {
        if (viewType == 0)
            return new CityHolder(inflateItemView(R.layout.item_wave_contact, parent));
        else
            return new PinnedHolder(inflateItemView(R.layout.item_pinned_header, parent));
    }


    @Override
    protected void convert(BaseViewHolder holder, ModelContactCity item) {
        if (holder instanceof CityHolder) {
            ((CityHolder) holder).city_name.setText(item.name);
        }else {
            String letter = item.pys.substring(0, 1);
            ((PinnedHolder) holder).city_tip.setText(letter);
        }
    }

    public int getLetterPosition(String letter){
        for (int i = 0 ; i < getData().size(); i++){
            if(getData().get(i).type ==1 && getData().get(i).pys.equals(letter)){
                return i;
            }
        }
        return -1;
    }

    class CityHolder extends BaseViewHolder {

        TextView city_name;

        public CityHolder(View view) {
            super(view);
            city_name = findViewById(R.id.tv_contact_name);
        }
    }


    class PinnedHolder extends BaseViewHolder {

        TextView city_tip;

        public PinnedHolder(View view) {
            super(view);
            city_tip = findViewById(R.id.city_tip);
        }
    }
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/wavesidebar/adapter/AdapterContactCity.kt
package com.tamsiree.rxui.view.wavesidebar.adapter

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.tamsiree.rxui.R
import com.tamsiree.rxui.model.ModelContactCity

/**
 * 此适配器仅供参考借鉴
 * @author tamsiree
 */
class AdapterContactCity : BaseWaveSideAdapter<ModelContactCity?, BaseViewHolder?> {
    constructor(context: Context?) : super(context)
    constructor(context: Context?, data: List<ModelContactCity?>?) : super(context, data)

    override fun getDefItemViewType(position: Int): Int {
        val city = getItem(position)
        return city!!.type
    }

    override fun onCreateDefViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        return if (viewType == 0) {
            CityHolder(inflateItemView(R.layout.item_wave_contact, parent))
        } else {
            PinnedHolder(inflateItemView(R.layout.item_pinned_header, parent))
        }
    }

    override fun convert(holder: BaseViewHolder?, item: ModelContactCity?) {
        if (holder is CityHolder) {
            holder.city_name!!.text = item?.name
        } else {
            val letter = item?.pys!!.substring(0, 1)
            (holder as PinnedHolder).city_tip!!.text = letter
        }
    }

    fun getLetterPosition(letter: String): Int {
        for (i in data.indices) {
            if (data[i]!!.type == 1 && data[i]!!.pys == letter) {
                return i
            }
        }
        return -1
    }

    internal inner class CityHolder(view: View?) : BaseViewHolder(view) {
        var city_name: TextView? = findViewById<TextView>(R.id.tv_contact_name)

    }

    internal inner class PinnedHolder(view: View?) : BaseViewHolder(view) {
        var city_tip: TextView? = findViewById<TextView>(R.id.city_tip)

    }
}